# Comencemos con la primera etapa: la importación de datos.

In [1]:
import pandas as pd

# Importar los conjuntos de datos
dataframeTrain = pd.read_csv("credit-train.csv")
dataframeTest = pd.read_csv("credit-test.csv")

# Mostrar las primeras filas y la información del conjunto de datos de entrenamiento
print("Primeras filas de dataframeTrain:")
print(dataframeTrain.head())
print("\nInformación del conjunto de datos de entrenamiento:")
print(dataframeTrain.info())

# Mostrar las primeras filas y la información del conjunto de datos de prueba
print("\nPrimeras filas de dataframeTest:")
print(dataframeTest.head())
print("\nInformación del conjunto de datos de prueba:")
print(dataframeTest.info())


Primeras filas de dataframeTrain:
   age          job  marital          education  default housing loan  \
0   44  blue-collar  married           basic.4y  unknown     yes   no   
1   53   technician  married            unknown       no      no   no   
2   28   management   single  university.degree       no     yes   no   
3   39     services  married        high.school       no      no   no   
4   55      retired  married           basic.4y       no     yes   no   

    contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
0  cellular   aug         thu  ...         1    999         0  nonexistent   
1  cellular   nov         fri  ...         1    999         0  nonexistent   
2  cellular   jun         thu  ...         3      6         2      success   
3  cellular   apr         fri  ...         2    999         0  nonexistent   
4  cellular   aug         fri  ...         1      3         1      success   

  emp_var_rate  cons_price_idx  cons_conf_idx  euribor3m  

 etapa 2 realizaremos un análisis numérico y visual de los datos. Comenzaremos por explorar las características numéricas en nuestros conjuntos de datos. Utilizaremos estadísticas descriptivas para comprender la distribución y las tendencias de estas características.

A continuación, analizaremos las características categóricas mediante gráficos de barras para comprender la distribución de las diferentes categorías en cada característica.

Comencemos con el análisis numérico.

Para el análisis numérico, calcularemos algunas estadísticas descriptivas para las características numéricas en nuestros conjuntos de datos, como la media, la mediana, la desviación estándar, el mínimo y el máximo. Esto nos dará una idea de la distribución y la dispersión de estos datos.

Veamos las estadísticas descriptivas para las características numéricas en el conjunto de datos de entrenamiento:

In [2]:
# Estadísticas descriptivas para características numéricas en el conjunto de datos de entrenamiento
numeric_features_train = dataframeTrain.select_dtypes(include=['int64', 'float64'])
numeric_descriptions_train = numeric_features_train.describe()
print("Estadísticas descriptivas para características numéricas en el conjunto de datos de entrenamiento:")
print(numeric_descriptions_train)


Estadísticas descriptivas para características numéricas en el conjunto de datos de entrenamiento:
               age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp_var_rate  cons_price_idx  cons_conf_idx     euribor3m  \
count  41188.000000    41188.000000   41188.000000  41188.000000   
mean       0.081886       93.575664     -40.502600      3.6212

# Las estadísticas descriptivas para las características numéricas en el conjunto de datos de entrenamiento son las siguientes:

age: La edad promedio de los clientes es de aproximadamente 40 años, con una desviación estándar de aproximadamente 10.42. La edad mínima es 17 y la máxima es 9
8.
duration: La duración promedio de las llamadas es de alrededor de 258 segundos, con una desviación estándar de aproximadamente 259. La duración mínima es 0 segundos y la máxima es 4918 segund
os.
campaign: El número promedio de contactos durante la campaña es de alrededor de 2.57, con una desviación estándar de aproximadamente 2.77. El número mínimo de contactos es 1 y el máximo es
 56.
pdays: El número promedio de días que han pasado desde el último contacto es de aproximadamente 962.48, con una desviación estándar de aproximadamente 186.91. El valor mínimo es 0 y el máximo es
 999.
previous: El número promedio de contactos anteriores a esta campaña es de aproximadamente 0.17, con una desviación estándar de aproximadamente 0.49. El valor mínimo es 0 y el máximo
 es 7.
emp_var_rate: La tasa de variación del empleo tiene una media de aproximadamente 0.08, con una desviación estándar de aproximadamente 1.57. El valor mínimo es -3.4 y el máximo 
es 1.4.
cons_price_idx: El índice de precios al consumidor tiene una media de aproximadamente 93.58, con una desviación estándar de aproximadamente 0.58. El valor mínimo es 92.201 y el máximo es
 94.767.
cons_conf_idx: El índice de confianza del consumidor tiene una media de aproximadamente -40.5, con una desviación estándar de aproximadamente 4.63. El valor mínimo es -50.8 y el máximo 
es -26.9.
euribor3m: La tasa Euribor a 3 meses tiene una media de aproximadamente 3.62, con una desviación estándar de aproximadamente 1.73. El valor mínimo es 0.634 y el máximo
 es 5.045.
nr_employed: El número de empleados tiene una media de aproximadamente 5167.04, con una desviación estándar de aproximadamente 72.25. El valor mínimo es 4963.6 y el máximo
 es 5228.1.
y: La columna objetivo tiene un promedio de alrededor de 0.11, lo que indica que aproximadamente el 11.27% de los clientes se suscribieron al crédito.

# Etapa 3: Análisis y Visualización de Datos
En esta etapa, realizaremos un análisis y visualización de los datos para comprender mejor las relaciones entre las características y la variable objetivo. Esto nos ayudará a identificar patrones y tendencias en los datos que pueden ser útiles para el modelado posterior.

Análisis de Variables Categóricas
Comencemos analizando las variables categóricas en nuestros conjuntos de datos. Calcularemos la frecuencia de cada categoría y visualizaremos estas distribuciones.

Vamos a proceder con este análisis.

Para analizar las variables categóricas en nuestros conjuntos de datos, podemos usar diferentes métodos dependiendo de la naturaleza de los datos. Algunas opciones incluyen:

Conteo de valores únicos: Podemos contar la cantidad de valores únicos para cada categoría en cada variable categórica.
Gráficos de barras: Podemos visualizar la distribución de las categorías utilizando gráficos de barras.
Tablas de contingencia: Podemos crear tablas de contingencia para analizar la relación entre diferentes variables categóricas.

In [3]:

# Identificar las columnas categóricas en el conjunto de datos de entrenamiento
categorical_cols_train = dataframeTrain.select_dtypes(include=['object']).columns.tolist()
print("Columnas categóricas en el conjunto de datos de entrenamiento:")
print(categorical_cols_train)

# Identificar las columnas categóricas en el conjunto de datos de prueba
categorical_cols_test = dataframeTest.select_dtypes(include=['object']).columns.tolist()
print("\nColumnas categóricas en el conjunto de datos de prueba:")
print(categorical_cols_test)


Columnas categóricas en el conjunto de datos de entrenamiento:
['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome']

Columnas categóricas en el conjunto de datos de prueba:
['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome']


hemos identificado las columnas categóricas en ambos conjuntos de datos, podemos decidir cómo codificarlas. Hay varias opciones, como la codificación one-hot, la codificación ordinal o la codificación basada en frecuencias.

Dado que todas las columnas categóricas tienen un número razonable de categorías y no hay un orden inherente en las categorías, podemos optar por la codificación one-hot.

Este código creará nuevas columnas para cada categoría en las columnas categóricas originales, utilizando la codificación one-hot. Luego, mostrará las primeras filas del conjunto de datos de entrenamiento codificado para que puedas verificar los resultados.

In [4]:
# Aplicar codificación one-hot a las columnas categóricas en el conjunto de datos de entrenamiento
dataframeTrain_encoded = pd.get_dummies(dataframeTrain, columns=['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome'])

# Mostrar las primeras filas del conjunto de datos de entrenamiento codificado
print("Conjunto de datos de entrenamiento codificado:")
print(dataframeTrain_encoded.head())


Conjunto de datos de entrenamiento codificado:
   age  duration  campaign  pdays  previous  emp_var_rate  cons_price_idx  \
0   44       210         1    999         0           1.4          93.444   
1   53       138         1    999         0          -0.1          93.200   
2   28       339         3      6         2          -1.7          94.055   
3   39       185         2    999         0          -1.8          93.075   
4   55       137         1      3         1          -2.9          92.201   

   cons_conf_idx  euribor3m  nr_employed  ...  month_oct  month_sep  \
0          -36.1      4.963       5228.1  ...      False      False   
1          -42.0      4.021       5195.8  ...      False      False   
2          -39.8      0.729       4991.6  ...      False      False   
3          -47.1      1.405       5099.1  ...      False      False   
4          -31.4      0.869       5076.2  ...      False      False   

   day_of_week_fri  day_of_week_mon  day_of_week_thu  day_of_we

# Codificación one-hot de las columnas categóricas en el conjunto de datos de prueba

In [5]:
# Importar la función necesaria
from sklearn.preprocessing import OneHotEncoder

# Crear un objeto OneHotEncoder
encoder = OneHotEncoder()

# Aplicar la codificación one-hot a las columnas categóricas en el conjunto de datos de prueba
dataframeTest_encoded = encoder.fit_transform(dataframeTest[categorical_cols])

# Convertir la matriz dispersa resultante a un DataFrame de pandas
dataframeTest_encoded = pd.DataFrame(dataframeTest_encoded.toarray(), columns=encoder.get_feature_names_out(categorical_cols))

# Concatenar el DataFrame codificado con el resto de las columnas del conjunto de datos de prueba
dataframeTest_encoded = pd.concat([dataframeTest.drop(columns=categorical_cols), dataframeTest_encoded], axis=1)

# Mostrar las primeras filas del conjunto de datos de prueba codificado
print("Conjunto de datos de prueba codificado:")
print(dataframeTest_encoded.head())


NameError: name 'categorical_cols' is not defined

In [ ]:
categorical_cols = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome']

dataframeTest_encoded = pd.get_dummies(dataframeTest, columns=categorical_cols)

# Mostrar las primeras filas del conjunto de datos de prueba codificado
print("Conjunto de datos de prueba codificado:")
print(dataframeTest_encoded.head())



# Calcular métricas de evaluación adicionales

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix

# Realiza predicciones en el conjunto de prueba
y_pred_test = modelo.predict(X_test)

# Calcular métricas de evaluación adicionales
report = classification_report(y_test, y_pred_test)
print("Reporte de clasificación:")
print(report)

# Calcular matriz de confusión
conf_matrix = confusion_matrix(y_test, y_pred_test)
print("Matriz de confusión:")
print(conf_matrix)


In [ ]:
from sklearn.metrics import classification_report, confusion_matrix

# Realiza predicciones en el conjunto de prueba
y_pred_test = modelo.predict(X_test)  # Reemplaza 'log_reg_model' con el nombre de tu modelo

# Calcular métricas de evaluación adicionales
report = classification_report(y_test, y_pred_test)
print("Reporte de clasificación:")
print(report)

# Calcular matriz de confusión
conf_matrix = confusion_matrix(y_test, y_pred_test)
print("Matriz de confusión:")
print(conf_matrix)


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Dividir los datos en características (X) y variable objetivo (y) para el conjunto de entrenamiento
X_train = dataframeTrain_encoded.drop(columns=['y'])
y_train = dataframeTrain_encoded['y']

# Dividir los datos en características (X) y variable objetivo (y) para el conjunto de prueba
X_test = dataframeTest_encoded
y_test = dataframeTest_encoded['y']  # Asignar solo la columna 'y'

# Dividir los datos de entrenamiento en conjuntos de entrenamiento y validación
X_train_split, X_val, y_train_split, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

# Inicializar el modelo de Regresión Logística
model = LogisticRegression()

# Ajustar el modelo a los datos de entrenamiento divididos
model.fit(X_train_split, y_train_split)

# Asegurarnos de que las características en X_test estén en el mismo orden que en X_train
X_test = X_test[X_train.columns]

# Predecir las etiquetas para los datos de prueba
y_pred_test = model.predict(X_test)

# Calcular la precisión del modelo en el conjunto de prueba
accuracy_test = accuracy_score(y_test, y_pred_test)
print("Precisión del modelo en el conjunto de prueba:", accuracy_test)


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Dividir los datos en características (X) y variable objetivo (y) para el conjunto de entrenamiento
X_train = dataframeTrain_encoded.drop(columns=['y'])
y_train = dataframeTrain_encoded['y']

# Dividir los datos en características (X) y variable objetivo (y) para el conjunto de prueba
X_test = dataframeTest_encoded.drop(columns=['y'])
y_test = dataframeTest_encoded['y']  # Asignar solo la columna 'y'

# Asegurarnos de que las columnas en X_train y X_test estén en el mismo orden
common_columns = X_train.columns.intersection(X_test.columns)
X_train = X_train.reindex(columns=common_columns)
X_test = X_test.reindex(columns=common_columns)

# Dividir los datos de entrenamiento en conjuntos de entrenamiento y validación
X_train_split, X_val, y_train_split, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

# Inicializar el modelo de Regresión Logística con un mayor número máximo de iteraciones
model = LogisticRegression(max_iter=1000)

# Ajustar el modelo a los datos de entrenamiento divididos
model.fit(X_train_split, y_train_split)

# Predecir las etiquetas para los datos de prueba
y_pred_test = model.predict(X_test)

# Convertir los valores de y_test a binarios
y_test_binary = y_test.replace({'yes': 1, 'no': 0})

# Calcular la precisión del modelo en el conjunto de prueba
accuracy_test = accuracy_score(y_test_binary, y_pred_test)
print("Precisión del modelo en el conjunto de prueba:", accuracy_test)


In [ ]:
# Filtrar y_test para eliminar las filas con valores 'unknown'
y_test_filtered = y_test[y_test.isin(['yes', 'no'])]

# Convertir los valores de y_test_filtered a binarios
y_test_binary = y_test_filtered.replace({'yes': 1, 'no': 0})

# Calcular la precisión del modelo en el conjunto de prueba
accuracy_test = accuracy_score(y_test_binary, y_pred_test)
print("Precisión del modelo en el conjunto de prueba:", accuracy_test)


## MODELO 1



In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Dividir los datos en características (X) y variable objetivo (y) para el conjunto de entrenamiento
X_train = dataframeTrain_encoded.drop(columns=['y'])
y_train = dataframeTrain_encoded['y']

# Dividir los datos de entrenamiento en conjuntos de entrenamiento y validación
X_train_split, X_val, y_train_split, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

# Inicializar el modelo de Regresión Logística con un mayor número máximo de iteraciones
model = LogisticRegression(max_iter=5000)

# Ajustar el modelo a los datos de entrenamiento divididos
model.fit(X_train_split, y_train_split)

# Predecir las etiquetas para los datos de validación
y_pred_val = model.predict(X_val)

# Calcular la precisión del modelo en el conjunto de validación
accuracy_val = accuracy_score(y_val, y_pred_val)
print("Precisión del modelo en el conjunto de validación:", accuracy_val)


coeficientes de las características ordenadas de mayor a menor importancia según su influencia en la predicción. Los coeficientes positivos indican que un aumento en esa característica aumenta la probabilidad de que la clase objetivo sea '1' (por ejemplo, 'sí'), mientras que los coeficientes negativos indican lo contrario.

In [ ]:
# Obtener los coeficientes del modelo
coefficients = pd.DataFrame(data=model.coef_.reshape(-1, 1), index=X_train.columns, columns=['Coefficient'])
coefficients.sort_values(by='Coefficient', ascending=False, inplace=True)

# Mostrar los coeficientes más importantes
print("Coeficientes más importantes:")
print(coefficients.head(10))


estos coeficientes nos proporcionan información valiosa sobre las características que más influyen en la predicción del modelo. Esta comprensión puede ser útil para mejorar aún más el modelo y optimizar las estrategias de marketing dirigidas a los clientes. 

month_mar (Mes de Marzo): Parece que el mes de marzo tiene una influencia significativa en la predicción positiva de la variable objetivo. Esto puede deberse a que marzo es un mes en el que se realizan acciones especiales o promociones que pueden aumentar la probabilidad de que un cliente acepte la oferta.

cons_price_idx (Índice de Precios al Consumidor): El índice de precios al consumidor también tiene un impacto positivo en la predicción. Esto sugiere que un mayor índice de precios al consumidor está asociado con una mayor probabilidad de que los clientes acepten la oferta.

default_no (No Incumplimiento de Pagos): La ausencia de incumplimiento de pagos (default_no) también tiene un efecto positivo en la predicción, lo cual es lógico ya que los clientes sin historial de incumplimiento son más propensos a aceptar la oferta.

contact_cellular (Contacto por Teléfono Celular): El contacto por teléfono celular también tiene una influencia positiva en la predicción, lo que sugiere que este método de contacto puede ser más efectivo para persuadir a los clientes.

job_student (Trabajo como Estudiante) y job_retired (Trabajo como Jubilado): Los trabajos como estudiante y jubilado también tienen una influencia positiva en la predicción, lo cual puede deberse a características específicas de estos grupos demográficos que los hacen más propensos a aceptar la preguntar.preguntar.

# realizar un análisis de residuos en el contexto de la Regresión Logística:

Obtención de las probabilidades pronosticadas: Utilizamos el modelo de Regresión Logística para predecir las probabilidades de pertenencia a cada clase para los datos de entrenamiento.

Cálculo de los residuos: Calculamos los residuos como la diferencia entre las probabilidades pronosticadas y las respuestas observadas (0 o 1 en el caso de la Regresión Logística).

Análisis de los residuos: Visualizamos los residuos para detectar patrones sistemáticos. Esto puede hacerse mediante un gráfico de residuos versus las características relevantes o mediante gráficos como el histograma de residuos.

Prueba de homocedasticidad: Comprobamos si la varianza de los residuos es constante en todo el rango de valores pronosticados. Esto puede hacerse mediante un gráfico de residuos versus los valores pronosticados.

Prueba de normalidad: Verificamos si los residuos siguen una distribución normal. Esto se puede hacer mediante un gráfico Q-Q (quantile-quantile) o una prueba estadística como la prueba de Shapiro-Wilk.

Identificación de valores atípicos: Buscamos valores atípicos o influencia excesiva mediante la identificación de observaciones con residuos altos.

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

# Suponiendo que "y_train_pred" son las probabilidades pronosticadas por el modelo
y_train_pred = model.predict_proba(X_train_split)[:, 1]  # Probabilidad de pertenencia a la clase positiva

# Calcular los residuos como la diferencia entre las probabilidades pronosticadas y las respuestas observadas
residuos = y_train_split - y_train_pred


In [ ]:
# Gráfico de dispersión de residuos versus valores pronosticados
plt.figure(figsize=(10, 6))
plt.scatter(y_train_pred, residuos, alpha=0.5)
plt.xlabel('Probabilidades pronosticadas')
plt.ylabel('Residuos')
plt.title('Gráfico de Residuos vs. Probabilidades Pronosticadas')
plt.axhline(y=0, color='r', linestyle='--')
plt.grid(True)
plt.show()


Este gráfico nos ayudará a identificar si hay alguna relación entre los residuos y las probabilidades pronosticadas. Queremos ver una dispersión aleatoria alrededor de la línea cero (la línea punteada roja). Si hay algún patrón, como una curva en forma de U o un patrón en zigzag, podría indicar que el modelo tiene problemas.

Hacer un histograma de los residuos para evaluar su distribución.

In [ ]:
# Histograma de los residuos
plt.figure(figsize=(10, 6))
plt.hist(residuos, bins=30, edgecolor='black')
plt.xlabel('Residuos')
plt.ylabel('Frecuencia')
plt.title('Histograma de Residuos')
plt.grid(True)
plt.show()


En este histograma, esperamos ver una distribución simétrica alrededor de cero, indicando que los residuos están distribuidos normalmente.

# explorar más a fondo la distribución de los residuos y realizar algunas pruebas estadísticas para evaluar su normalidad.

Primero, vamos a importar las bibliotecas necesarias: 
import seaborn as sns
from scipy import stats

In [ ]:
import seaborn as sns
from scipy import stats

# Eliminar columnas no numéricas si es necesario


In [ ]:
# Eliminar columnas no numéricas si es necesario
dataframeTrain_numeric = dataframeTrain_encoded.select_dtypes(include=[np.number])

# Identificar valores atípicos usando el método de Tukey
Q1 = dataframeTrain_numeric.quantile(0.25)
Q3 = dataframeTrain_numeric.quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

outliers = ((dataframeTrain_numeric < lower_bound) | (dataframeTrain_numeric > upper_bound)).any(axis=1)

# Eliminar valores atípicos
dataframeTrain_clean = dataframeTrain_encoded[~outliers]



# Prueba de normalidad (Shapiro-Wilk)


In [ ]:

stat, p_value = stats.shapiro(residuos)
print("Estadístico de la prueba:", stat)
print("Valor p:", p_value)

alpha = 0.05
if p_value > alpha:
    print("No se rechaza la hipótesis nula: los residuos parecen provenir de una distribución normal.")
else:
    print("Se rechaza la hipótesis nula: los residuos no parecen provenir de una distribución normal.")


Si el valor p es mayor que el nivel de significancia (generalmente 0.05), no podemos rechazar la hipótesis nula y podemos decir que los residuos se distribuyen normalmente. De lo contrario, si el valor p es menor que el nivel de significancia, rechazamos la hipótesis nula y concluimos que los residuos no se distribuyen normalmente.

Este análisis nos proporciona información importante sobre la calidad del modelo y puede ayudarnos a identificar áreas de mejora. Si los residuos no se distribuyen normalmente o muestran algún patrón sistemático, podríamos necesitar revisar el modelo y considerar ajustes adicionales.

In [ ]:
from sklearn.model_selection import GridSearchCV

# Definir la cuadrícula de valores para los hiperparámetros
param_grid = {
    'penalty': ['l1', 'l2'],  # Tipo de penalización (L1 o L2)
    'C': [0.001, 0.01, 0.1, 1, 10, 100]  # Parámetro de regularización
}

# Inicializar el modelo de Regresión Logística
model = LogisticRegression(max_iter=100000)

# Inicializar el GridSearchCV con el modelo y la cuadrícula de parámetros
grid_search = GridSearchCV(model, param_grid, cv=5, scoring='accuracy')

# Ajustar el GridSearchCV a los datos de entrenamiento
grid_search.fit(X_train_split, y_train_split)

# Mostrar los mejores hiperparámetros encontrados
print("Mejores hiperparámetros encontrados:")
print(grid_search.best_params_)

# Mostrar la mejor puntuación (exactitud) obtenida durante la búsqueda
print("Mejor puntuación (exactitud):", grid_search.best_score_)


In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

# Definir la cuadrícula de valores para los hiperparámetros
param_grid = {
    'penalty': ['l2'],  # Cambiamos 'l1' por 'l2'
    'C': [0.001, 0.01, 0.1, 1, 10, 100]  # Parámetro de regularización
}

# Inicializar el modelo de Regresión Logística
model = LogisticRegression(max_iter=100000, solver='lbfgs')  # Especificamos el solver

# Inicializar el GridSearchCV con el modelo y la cuadrícula de parámetros
grid_search = GridSearchCV(model, param_grid, cv=5, scoring='accuracy')

# Ajustar el GridSearchCV a los datos de entrenamiento
grid_search.fit(X_train_split, y_train_split)

# Mostrar los mejores hiperparámetros encontrados
print("Mejores hiperparámetros encontrados:")
print(grid_search.best_params_)

# Mostrar la mejor puntuación (exactitud) obtenida durante la búsqueda
print("Mejor puntuación (exactitud):", grid_search.best_score_)


encontrado los mejores hiperparámetros para tu modelo de Regresión Logística:

Valor de C: 1
Tipo de penalización: L2 (Ridge)
La mejor puntuación de exactitud alcanzada con estos hiperparámetros es de aproximadamente 0.911, lo que indica un buen rendimiento del modelo en el conjunto de validación.

Ahora que tienes estos mejores hiperparámetros, puedes utilizarlos para ajustar tu modelo de Regresión Logística y luego evaluar su desempeño en el conjunto de prueba para asegurarte de que generaliza bien a datos nuevos y no vistos.

# Inicializar un nuevo modelo de Regresión Logística con los mejores hiperparámetros encontrados

In [ ]:

best_model = LogisticRegression(max_iter=100000, penalty=grid_search.best_params_['penalty'], 
                                C=grid_search.best_params_['C'])

# Entrenar el modelo con los datos de entrenamiento completos (no divididos)
best_model.fit(X_train, y_train)

# Predecir las etiquetas para el conjunto de prueba
y_pred_test = best_model.predict(X_test)

# Calcular la precisión del modelo en el conjunto de prueba
accuracy_test = accuracy_score(y_test, y_pred_test)
print("Precisión del modelo en el conjunto de prueba:", accuracy_test)



# paso 1: predecir las etiquetas en el conjunto de prueba utilizando el modelo ajustado con los mejores hiperparámetros.

# Matriz de Confusión:

In [ ]:
# Por ejemplo, si tus características están en un DataFrame llamado dataframeTrain_clean y tus etiquetas en una Serie llamada y:
X = dataframeTrain_clean.drop(columns=['y'])
y = dataframeTrain_clean['y']


In [ ]:
# Suponiendo que tienes un modelo llamado 'modelo' ya entrenado
# y deseas obtener las predicciones para el conjunto de prueba 'X_test'
y_pred = model.predict(X_test)



In [ ]:
from sklearn.metrics import confusion_matrix
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
X = dataframeTrain_clean.drop(columns=['y'])
y = dataframeTrain_clean['y']
# Suponiendo que tienes tus características en X y tus etiquetas en y
# Reemplaza X e y con tus propios datos
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Ahora puedes utilizar X_train, X_test, y_train, y y_test para entrenar tu modelo y evaluar su rendimiento.

# Aquí se supone que tienes tus predicciones (y_pred) y tus etiquetas verdaderas (y_test).
y_pred = model.predict(X_test)
# Si ya tienes y_pred y y_test definidos, puedes calcular la matriz de confusión y visualizarla.

# Calcular la matriz de confusión
cm = confusion_matrix(y_test, y_pred)

# Visualizar la matriz de confusión utilizando seaborn
plt.figure(figsize=(8, 6))
sns.heatmap(cm, annot=True, fmt="d", cmap="Blues", cbar=False)
plt.xlabel("Predicted labels")
plt.ylabel("True labels")
plt.title("Confusion Matrix")
plt.show()


# Curva ROC y AUC:


In [ ]:
from sklearn.metrics import roc_curve, roc_auc_score

# Calcular las probabilidades de las clases positivas
y_prob = best_model.predict_proba(X_test)[:, 1]

# Calcular la tasa de verdaderos positivos y la tasa de falsos positivos
fpr, tpr, thresholds = roc_curve(y_test, y_prob)

# Calcular el área bajo la curva ROC (AUC)
auc = roc_auc_score(y_test, y_prob)

# Visualizar la curva ROC
plt.figure(figsize=(8, 6))
plt.plot(fpr, tpr, color='blue', label=f'ROC Curve (AUC = {auc:.2f})')
plt.plot([0, 1], [0, 1], color='red', linestyle='--')
plt.xlabel('Tasa de Falsos Positivos')
plt.ylabel('Tasa de Verdaderos Positivos')
plt.title('Curva ROC')
plt.legend()
plt.show()


In [ ]:
from sklearn.metrics import precision_recall_curve, auc

# Calcular la precisión y el recall
precision, recall, _ = precision_recall_curve(y_test, y_prob)

# Calcular el área bajo la curva PR (AUC)
pr_auc = auc(recall, precision)

# Visualizar la curva de Precisión-Recall
plt.figure(figsize=(8, 6))
plt.plot(recall, precision, color='blue', label=f'PR Curve (AUC = {pr_auc:.2f})')
plt.xlabel('Recall')
plt.ylabel('Precision')
plt.title('Curva de Precisión-Recall')
plt.legend()
plt.show()


In [ ]:
# Obtener los coeficientes de la regresión logística
coefficients = best_model.coef_[0]
feature_names = X_train.columns

# Visualizar los coeficientes
plt.figure(figsize=(10, 6))
plt.barh(feature_names, coefficients, color='skyblue')
plt.xlabel('Coeficiente')
plt.ylabel('Característica')
plt.title('Coeficientes de Regresión Logística')
plt.show()


In [ ]:
if hasattr(best_model, 'feature_importances_'):
    # Obtener la importancia de las características
    importances = best_model.feature_importances_
    indices = np.argsort(importances)[::-1]

    # Visualizar la importancia de las características
    plt.figure(figsize=(10, 6))
    plt.barh(feature_names[indices], importances[indices], color='green')
    plt.xlabel('Importancia')
    plt.ylabel('Característica')
    plt.title('Importancia de Características (Random Forest)')
    plt.show()
else:
    print("El modelo no tiene atributo 'feature_importances_'")
